standalone/pruntime/enclave/src/msg/msg_trait.rs

```diff
@@ -0,0 +1,9 @@
+pub trait MsgTrait<A,B,C,D,E> {
+    fn push(&mut self,a:A);
+    fn received(&mut self,b:B);
+    fn get_msgs(&mut self,c:C)->D;
+    fn get_sequence(&mut self)->E;
```

Comment:
Let's assume sequence ids are all `u64`. It's not necessary to define every parameter as generic types. IMO a reasonable definition could be:

```rust
trait MsgChannel<M: Encode + Decode> {  // M can be serialized by parity-scale-codec
    fn push(&mut self, msg: &M);
    fn received(&mut self, seq: u64);  // Called when the receiver confirms the acceptance of a message
    fn get_sequence(&self);  // It's a const function, right?
}
```

Reply:
To make it event better, I think it makes more sense to define a message channel as a struct, rather than a trait, because all the message channels have very similar functionalities:

1. The sender can push some message into the channel queue:
    - The message should have a monotone increment `sequence_id`
    - The message will be signed by some key (for System contract, signed by Identity Key; for other contracts, signed by the contract key... but actually now it is the same as Identity Key)
    - The message will stay in the queue, waiting for other to query
2. The queue can be queried with `seq_id >= n` (e.g. the relayer "phost")
3. The receiver confirm the acceptance of a message (by sequence id):
    - This happens when the TEE noticed a blockchain event indicating some sequence_id is processed
    - The blockchain guarantee the `sequenced_id` will be contiguous (no missing message)
    - Then the the channel can remove the received message from the queue

Let's carefully define these structures, and avoid to write duplicated code. Some tips:

1. The message channel should contain a queue as a buffer of the unconfirmed messages
2. The message in the queue must have the message body (`Encode + Decode` so we can serialize it), a `sequence_id`, and a cryptographic signature

Reply:
Probably we can make a trait for messages:

```rust
trait Message: Encode + Decode {}
struct SignedMessage<M: Message> {
    body: M,
    seq: u64,
    sig: <...>,
}

impl<M: Message> SignedMessage {
    fn new(body: M, seq: u64, key: &Pair) -> Self {
        Self {
            body,
            seq,
            sig: Self::sign(body, seq)
        }
    }
    fn sign(body: M, seq: u64, key: &Pair) -> <...> {
        let sighash = blake256(Encode::encode((body, seq)));
        key.sign(sighash)
    }
    fn verify(&self) -> bool { /* ... */ }
}
```

Reply:
I see you have `struct MsgTunnel` serving as the above purpose, but it saves `Vec<u8>` with the message type defined as an enum. However you still need to `impl MsgTrait<...>` in each contract, and do the conversion time to time.

If you go with the above approach, I think we can just define a `MsgChannel<BalanceTranser>` in `Balances`, and `MsgChannel<AssetsTransfer>` in `Assets`, without the need to impl a new struct?

Then you can feel safe to define the `MsgChannel` inside the contract struct itself like this (or you can also put them somewhere outside the contract, like what you did currently):

```rust
struct Balances {
    // ...
    channel: MsgChannel<BalanceTransfer>,
}

impl Balances {
    // ...
    fn foobar(...) {
        self.channel.push(...);
        self.channel.received(...);
    }
}
```